## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

### Review of data types in df

In [3]:
for column in attrition_df.columns:
  if type(attrition_df[column][0])==str:
    print('*'*5,column,'*'*5)
    print(attrition_df[column].value_counts())
    print()

***** Attrition *****
No     1233
Yes     237
Name: Attrition, dtype: int64

***** BusinessTravel *****
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: BusinessTravel, dtype: int64

***** Department *****
Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64

***** EducationField *****
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: EducationField, dtype: int64

***** JobRole *****
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Director             80
Human Resources               52
Name: JobRole, dtype: int64

***** MaritalStatus *****
Married     673
Single      470
Divorced    327
Name: Mari

In [4]:
attrition_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 27 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   Department                1470 non-null   object
 4   DistanceFromHome          1470 non-null   int64 
 5   Education                 1470 non-null   int64 
 6   EducationField            1470 non-null   object
 7   EnvironmentSatisfaction   1470 non-null   int64 
 8   HourlyRate                1470 non-null   int64 
 9   JobInvolvement            1470 non-null   int64 
 10  JobLevel                  1470 non-null   int64 
 11  JobRole                   1470 non-null   object
 12  JobSatisfaction           1470 non-null   int64 
 13  MaritalStatus             1470 non-null   object
 14  NumCompaniesWorked      

### Back to Original Instruction Cells

In [5]:
# Create y_df with the Attrition and Department columns
y_df=attrition_df[['Attrition','Department']]
y_df

Attrition              Department
0          Yes                   Sales
1           No  Research & Development
2          Yes  Research & Development
3           No  Research & Development
4           No  Research & Development
...        ...                     ...
1465        No  Research & Development
1466        No  Research & Development
1467        No  Research & Development
1468        No                   Sales
1469        No  Research & Development

[1470 rows x 2 columns]

In [6]:
attrition_df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [7]:
# Create a list of at least 10 column names to use as X data
X=attrition_df.drop(columns=['Attrition','Department'])


# Create X_df using your selected columns
X_df=pd.DataFrame(X)

# Show the data types for X_df
X_df.dtypes


Age                          int64
BusinessTravel              object
DistanceFromHome             int64
Education                    int64
EducationField              object
EnvironmentSatisfaction      int64
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
NumCompaniesWorked           int64
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StockOptionLevel             int64
TotalWorkingYears            int64
TrainingTimesLastYear        int64
WorkLifeBalance              int64
YearsAtCompany               int64
YearsInCurrentRole           int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
dtype: object

### Lazy_Encoder

In [8]:
def lazy_Encoder(df):
  from sklearn.preprocessing import OneHotEncoder
  ohe=OneHotEncoder(handle_unknown='ignore', sparse_output=False).set_output(transform='pandas')
  for column in df.columns:
    if df[column].dtype=='O':
      if len(df[column].value_counts())==2:
        new_cols=pd.get_dummies(df[column], dtype=int, drop_first=True)
        df[column]=new_cols
      else :
        new_cols=ohe.fit_transform(df[[column]])
        df=pd.concat([df,new_cols], axis=1).drop(columns=[column])
  return df

### Back to Original Instruction Cells

In [9]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_attrition_train, y_attrition_test, y_department_train, y_department_test\
 =train_test_split(X_df,y_df['Attrition'],y_df['Department'],random_state=42)

In [10]:
y_department_train

1343    Research & Development
1121                     Sales
1048                     Sales
1393                     Sales
527                      Sales
                 ...          
1130    Research & Development
1294    Research & Development
860     Research & Development
1459    Research & Development
1126                     Sales
Name: Department, Length: 1102, dtype: object

In [11]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
X_train_enc=lazy_Encoder(X_train)
X_test_enc=lazy_Encoder(X_test)


In [12]:
# Create a StandardScaler
ss_model=StandardScaler()

# Fit the StandardScaler to the training data
ss_model.fit(X_train_enc)

# Scale the training and testing data
X_train_scaled_df=pd.DataFrame(ss_model.transform(X_train_enc))
X_test_scaled_df=pd.DataFrame(ss_model.transform(X_test_enc))


In [13]:
y_department_train

1343    Research & Development
1121                     Sales
1048                     Sales
1393                     Sales
527                      Sales
                 ...          
1130    Research & Development
1294    Research & Development
860     Research & Development
1459    Research & Development
1126                     Sales
Name: Department, Length: 1102, dtype: object

In [14]:
# Create a OneHotEncoder for the Department column
from sklearn.preprocessing import OneHotEncoder
ohe=OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
ohe.fit(np.array(y_department_train).reshape(-1,1))


# Create two new variables by applying the encoder
# to the training and testing data
y_Department_train_enc=ohe.transform(np.array(y_department_train).reshape(-1,1))
y_Department_test_enc=ohe.transform(np.array(y_department_test).reshape(-1,1))

In [15]:
# Create a OneHotEncoder for the Attrition column
ohe=OneHotEncoder(sparse_output=False)


# Fit the encoder to the training data
ohe.fit(np.array(y_attrition_train).reshape(-1,1))

# Create two new variables by applying the encoder
# to the training and testing data
y_Attrition_train_enc=ohe.transform(np.array(y_attrition_train).reshape(-1,1))
y_Attrition_test_enc=ohe.transform(np.array(y_attrition_test).reshape(-1,1))

## Create, Compile, and Train the Model

In [31]:
# Find the number of columns in the X training data
input_nodes=len(X_train_scaled_df.columns)

# Create the input layer
input_layer= layers.Input(shape=input_nodes,name='input_features')

# Create at least two shared layers
shared_layer1=layers.Dense(64,activation='relu', name='shared_layer1')(input_layer)
shared_layer2=layers.Dense(128,activation='relu', name='shared_layer2')(shared_layer1)
shared_layer3=layers.Dense(64,activation='relu', name='shared_layer3')(shared_layer2)

In [32]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
shared_layer4=layers.Dense(32,activation='relu',name='shared_layer4')(shared_layer3)

# Create the output layer
Department_output = layers.Dense(3,activation='sigmoid',name='Department_output')(shared_layer4)


In [33]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
shared_layer5=layers.Dense(32,activation='relu',name='shared_layer5')(shared_layer3)

# Create the output layer
Attrition_output = layers.Dense(2,activation='sigmoid',name='Attrition_output')(shared_layer4)


In [34]:
# Create the model
model=Model(inputs=input_layer,outputs=[Department_output, Attrition_output])

# Compile the model
model.compile(optimizer='adam',
              loss={'Department_output':'binary_crossentropy',
                    'Attrition_output':'binary_crossentropy'},
              metrics= {'Department_output':'accuracy',
                    'Attrition_output':'accuracy'}
              )

# Summarize the model
model.summary()

Model: "model_3"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_features (InputLayer  [(None, 42)]                 0         []                            
 )                                                                                                
                                                                                                  
 shared_layer1 (Dense)       (None, 64)                   2752      ['input_features[0][0]']      
                                                                                                  
 shared_layer2 (Dense)       (None, 128)                  8320      ['shared_layer1[0][0]']       
                                                                                                  
 shared_layer3 (Dense)       (None, 64)                   8256      ['shared_layer2[0][0]'] 

In [35]:
# Train the model
model.fit(X_train_scaled_df,
          {'Department_output': y_Department_train_enc,
           'Attrition_output': y_Attrition_train_enc},
          epochs=100,
          batch_size=32,
          validation_split=0.2
          )

Epoch 1/100
28/28 [==============================] - 2s 18ms/step - loss: 1.2746 - Department_output_loss: 0.5948 - Attrition_output_loss: 0.6798 - Department_output_accuracy: 0.6220 - Attrition_output_accuracy: 0.5812 - val_loss: 1.0472 - val_Department_output_loss: 0.5120 - val_Attrition_output_loss: 0.5352 - val_Department_output_accuracy: 0.6063 - val_Attrition_output_accuracy: 0.7873
Epoch 2/100
28/28 [==============================] - 0s 4ms/step - loss: 0.8057 - Department_output_loss: 0.3629 - Attrition_output_loss: 0.4427 - Department_output_accuracy: 0.7582 - Attrition_output_accuracy: 0.8388 - val_loss: 0.7545 - val_Department_output_loss: 0.2760 - val_Attrition_output_loss: 0.4785 - val_Department_output_accuracy: 0.8914 - val_Attrition_output_accuracy: 0.7873
Epoch 3/100
28/28 [==============================] - 0s 4ms/step - loss: 0.5559 - Department_output_loss: 0.1569 - Attrition_output_loss: 0.3990 - Department_output_accuracy: 0.9251 - Attrition_output_accuracy: 0.8388

In [36]:
# Evaluate the model with the testing data
test_results=model.evaluate(X_test_enc,
 {'Department_output': y_Department_test_enc, 'Attrition_output': y_Attrition_test_enc})
test_results

12/12 [==============================] - 0s 2ms/step - loss: 21.3503 - Department_output_loss: 13.5286 - Attrition_output_loss: 7.8217 - Department_output_accuracy: 0.5924 - Attrition_output_accuracy: 0.8288


[21.350284576416016,
 13.528629302978516,
 7.821651935577393,
 0.592391312122345,
 0.8288043737411499]

In [37]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy : {test_results[-2]}")
print(f"Attrition Accuracy : {test_results[-1]}")

Department Accuracy : 0.592391312122345
Attrition Accuracy : 0.8288043737411499


With 2 core shared layers we got :
- Department Accuracy : 0.573369562625885
- Attrition Accuracy : 0.633152186870575

By adding a 3rd core layer we got :
- Department Accuracy : 0.592391312122345
- Attrition Accuracy : 0.8288043737411499

# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. **Yes**, because we are looking for a match or no match result
2. **sigmoid**, the results to compare to are 0 and 1, sigmoid gives 0 to 1 results that get rounded to 0 or 1
3. Add more layers and/or mode nodes to the result with the lowest accuracy